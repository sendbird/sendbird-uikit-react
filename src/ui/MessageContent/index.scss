@import '../../styles/variables';

:root {
  --sendbird-feedback-buttons-container-height: 40px;
  --sendbird-feedback-buttons-container-height-margin-bottom: 4px;
}

.sendbird-message-content {
  display: inline-flex;
  flex-direction: row;
  position: relative;
  width: 100%;

  &.incoming {
    justify-content: flex-start;
  }

  &.outgoing {
    justify-content: flex-end;
  }

  .sendbird-message-content__middle {
    max-width: 400px;
    @include mobile() {
      max-width: calc(100vw - 140px);
    }
  }

  .sendbird-message-content__middle {
    .sendbird-message-content__middle__quote-message.use-quote {
      bottom: -8px;
      margin-top: -8px;
    }
  }
}

.sendbird-message-content__feedback {
  margin-bottom: calc(var(--sendbird-feedback-buttons-container-height) + var(--sendbird-feedback-buttons-container-height-margin-bottom));
}

.sendbird-message-content__middle__quote-message__quote {
  width: 100%;
}

// incoming
.sendbird-message-content.incoming {
  .sendbird-message-content__left {
    display: inline-flex;
    min-width: 40px;
    position: relative;

    .sendbird-message-content__left__avatar {
      bottom: 2px;
      left: 0;
      position: absolute;
      &.multiple {
        border: 2px solid;
        @include themed() {
          border-color: t(bg-0);
        }
      }
    }

    .sendbird-message-content__left__avatar.use-thread-replies {
      bottom: 35px;
    }
  }

  .sendbird-message-content__middle {
    align-items: flex-start;
    display: inline-flex;
    flex-direction: column;
    position: relative;

    .sendbird-message-content__middle__body-container {
      .sendbird-message-content__middle__body-container__created-at {
        bottom: 6px;
        display: flex;
        flex-direction: row;
        min-width: 80px;
        position: absolute;
        right: -84px;
        white-space: nowrap;

        &.sendbird-mouse-hover {
          display: none;
        }
      }
    }

    .sendbird-message-content__middle__sender-name {
      margin-bottom: 4px;
      margin-left: 12px;
      overflow-x: hidden;
      position: relative;
      text-overflow: ellipsis;
      width: 100%;
    }

    .sendbird-message-content__middle__quote-message {
      display: inline-flex;
      position: relative;
      width: 100%;
      &.outgoing { justify-content: flex-end; }
      &.incoming { justify-content: flex-start; }

      &:hover {
        cursor: pointer;
      }
    }
  }

  .sendbird-message-content__right {
    display: inline-flex;
    margin-left: 4px;
    padding-top: 18px;
    position: relative;
    width: 50px;
    &.use-reactions { width: 70px; }

    &.chain-top {
      padding-top: 2px;

      &.use-quote {
        padding-top: 18px;
      }
    }

    .sendbird-message-content-menu {
      display: none;
      flex-direction: row;
      height: 32px;
      position: relative;

      &.sendbird-mouse-hover {
        display: inline-flex;
      }
    }
  }

  &:hover {
    @include mobile() {
      .sendbird-message-content__middle {
        .sendbird-message-content__middle__body-container {
          .sendbird-message-content__middle__body-container__created-at {
            display: inline-flex;
          }
        }
      }
    }

    .sendbird-message-content__right {
      .sendbird-message-content-menu {
        display: inline-flex;
      }
    }

    .sendbird-message-content__middle {
      .sendbird-message-content__middle__body-container {
        .sendbird-message-content__middle__body-container__created-at {
          display: none;
        }
      }
    }
  }
}

// outgoing
.sendbird-message-content.outgoing {
  .sendbird-message-content__left {
    box-sizing: border-box;
    display: inline-flex;
    justify-content: flex-end;
    position: relative;
    width: 50px;
    &.use-reactions { width: 70px; }

    &.use-quote {
      .sendbird-message-content-menu {
        top: 18px;
      }
    }

    .sendbird-message-content-menu {
      display: none;
      flex-direction: row;
      position: absolute;
      right: 4px;
      top: 2px;

      &.sendbird-mouse-hover {
        display: inline-flex;
      }
    }
  }

  .sendbird-message-content__middle {
    align-items: flex-end;
    display: inline-flex;
    flex-direction: column;
    position: relative;

    .sendbird-message-content__middle__quote-message {
      display: inline-flex;
      position: relative;
      width: 100%;
      &.outgoing { justify-content: flex-end; }
      &.incoming { justify-content: flex-start; }
    }

    .sendbird-message-content__middle__body-container {
      position: relative;
      width: 100%;

      .sendbird-message-content__middle__body-container__created-at {
        bottom: 2px;
        box-sizing: content-box;
        display: flex;
        justify-content: flex-end;
        left: -84px;
        min-height: 16px;
        min-width: 80px;
        position: absolute;
        white-space: nowrap;

        &.sendbird-mouse-hover {
          display: none;
        }

        .sendbird-message-content__middle__body-container__created-at__component-container {
          display: inline-flex;
          position: relative;
        }
      }
    }
  }

  .sendbird-message-content__right {
    display: none;
  }

  &:hover {
    @include mobile {
      .sendbird-message-content__middle {
        .sendbird-message-content__middle__body-container {
          .sendbird-message-content__middle__body-container__created-at {
            display: inline-flex;
          }
        }
      }
    }

    .sendbird-message-content__left {
      .sendbird-message-content-menu {
        display: inline-flex;
      }
    }

    .sendbird-message-content__middle {
      .sendbird-message-content__middle__body-container {
        .sendbird-message-content__middle__body-container__created-at {
          display: none;
        }
      }
    }
  }
}

.sendbird-message-content__middle__body-container {
  display: flex;
  flex-direction: column;
  position: relative;
  width: fit-content;

  .sendbird-message-content__middle__message-item-body {
    box-sizing: border-box;
    width: 100%;
  }
}

.sendbird-message-content-reactions {
  border-radius: 0 0 16px 16px;
  max-width: 400px;
  position: relative;
  width: 100%;
  @include themed() { background-color: t(bg-1); }

  &.primary {
    @include themed() { background-color: t(primary-3); }
  }

  &.mouse-hover,
  &:hover {
    @include themed() { background-color: t(bg-2); }

    &.primary {
      @include themed() { background-color: t(primary-4); }
    }
  }

  &.image-grid {
    @include themed() { background-color: t(bg-1); }
  }
}

// threads
.sendbird-message-content__middle__thread-replies {
  margin-top: 4px;
}

.sendbird-voice-message-item-body.sendbird-message-content__middle__message-item-body {
  height: 50px;
  min-width: 136px;
}

.sendbird-message-content__middle__body-container__feedback-buttons-container {
  position: absolute;
  top: var(--sendbird-feedback-buttons-container-height);
  display: flex;
  gap: 4px;
}