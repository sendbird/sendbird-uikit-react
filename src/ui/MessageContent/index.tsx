import React, {
  ReactElement,
  useContext,
  useRef,
  useState,
} from 'react';
import format from 'date-fns/format';
import './index.scss';

import Avatar from '../Avatar';
import UserProfile from '../UserProfile';
import MessageStatus from '../MessageStatus';
import MessageItemMenu from '../MessageItemMenu';
import MessageItemReactionMenu from '../MessageItemReactionMenu';
import ContextMenu, { MenuItems } from '../ContextMenu';
import Label, { LabelTypography, LabelColors } from '../Label';
import EmojiReactions from '../EmojiReactions';

import ClientAdminMessage from '../AdminMessage';
import TextMessageItemBody from '../TextMessageItemBody';
import FileMessageItemBody from '../FileMessageItemBody';
import ThumbnailMessageItemBody from '../ThumbnailMessageItemBody';
import OGMessageItemBody from '../OGMessageItemBody';
import UnknownMessageItemBody from '../UnknownMessageItemBody';
import QuoteMessage from '../QuoteMessage';

import {
  getClassName,
  getUIKitMessageTypes,
  getUIKitMessageType,
  isTextMessage,
  isOGMessage,
  isThumbnailMessage,
  getSenderName,
  isSentMessage,
  isVoiceMessage,
} from '../../utils';
import { UserProfileContext } from '../../lib/UserProfileContext';
import { ReplyType } from '../../index';
import { useLocalization } from '../../lib/LocalizationContext';
import useSendbirdStateContext from '../../hooks/useSendbirdStateContext';
import { GroupChannel } from '@sendbird/chat/groupChannel';
import { EmojiContainer } from '@sendbird/chat';
import { AdminMessage, FileMessage, Sender, UserMessage } from '@sendbird/chat/message';
import useLongPress from '../../hooks/useLongPress';
import MobileMenu from '../MobileMenu';
import { useMediaQueryContext } from '../../lib/MediaQueryContext';
import ThreadReplies from '../ThreadReplies';
import { ThreadReplySelectType } from '../../modules/Channel/context/const';
import VoiceMessageItemBody from '../VoiceMessageItemBody';
import { Nullable } from '../../types';

// should initialize in UserProfileContext.jsx
export interface UserProfileContextInterface {
  disableUserProfile: boolean;
  isOpenChannel: boolean;
  renderUserProfile?: (props: { user: Sender, close: () => void }) => React.ReactElement,
}

interface Props {
  className?: string | Array<string>;
  userId: string;
  channel: Nullable<GroupChannel>;
  message: AdminMessage | UserMessage | FileMessage;
  disabled?: boolean;
  chainTop?: boolean;
  chainBottom?: boolean;
  isReactionEnabled?: boolean;
  disableQuoteMessage?: boolean;
  replyType?: ReplyType;
  threadReplySelectType?: ThreadReplySelectType;
  nicknamesMap?: Map<string, string>;
  emojiContainer?: EmojiContainer;
  scrollToMessage?: (createdAt: number, messageId: number) => void;
  showEdit?: (bool: boolean) => void;
  showRemove?: (bool: boolean) => void;
  showFileViewer?: (bool: boolean) => void;
  resendMessage?: (message: UserMessage | FileMessage) => Promise<UserMessage | FileMessage>;
  toggleReaction?: (message: UserMessage | FileMessage, reactionKey: string, isReacted: boolean) => void;
  setQuoteMessage?: (message: UserMessage | FileMessage) => void;
  onReplyInThread?: (props: { message: UserMessage | FileMessage }) => void;
  onQuoteMessageClick?: (props: { message: UserMessage | FileMessage }) => void;
  onMessageHeightChange?: () => void;
}
export default function MessageContent({
  className,
  userId,
  channel,
  message,
  disabled = false,
  chainTop = false,
  chainBottom = false,
  isReactionEnabled = false,
  disableQuoteMessage = false,
  replyType,
  threadReplySelectType,
  nicknamesMap,
  emojiContainer,
  scrollToMessage,
  showEdit,
  showRemove,
  showFileViewer,
  resendMessage,
  toggleReaction,
  setQuoteMessage,
  onReplyInThread,
  onQuoteMessageClick,
  onMessageHeightChange,
}: Props): ReactElement {
  const messageTypes = getUIKitMessageTypes();
  const { dateLocale } = useLocalization();
  const { config } = useSendbirdStateContext?.() || {};
  const { disableUserProfile, renderUserProfile }: UserProfileContextInterface = useContext(UserProfileContext);
  const avatarRef = useRef(null);
  const contentRef = useRef(null);
  const { isMobile } = useMediaQueryContext();
  const [showMenu, setShowMenu] = useState(false);
  const [mouseHover, setMouseHover] = useState(false);
  const [supposedHover, setSupposedHover] = useState(false);

  const isByMe = (userId === (message as UserMessage | FileMessage)?.sender?.userId)
    || ((message as UserMessage | FileMessage)?.sendingStatus === 'pending')
    || ((message as UserMessage | FileMessage)?.sendingStatus === 'failed');
  const isByMeClassName = isByMe ? 'outgoing' : 'incoming';
  const chainTopClassName = chainTop ? 'chain-top' : '';
  const isReactionEnabledInChannel = isReactionEnabled && !channel?.isEphemeral;
  const isReactionEnabledClassName = isReactionEnabledInChannel ? 'use-reactions' : '';
  const supposedHoverClassName = supposedHover ? 'sendbird-mouse-hover' : '';
  const useReplying = !!((replyType === 'QUOTE_REPLY' || replyType === 'THREAD')
    && message?.parentMessageId && message?.parentMessage
    && !disableQuoteMessage
  );
  const useReplyingClassName = useReplying ? 'use-quote' : '';

  // Thread replies
  const displayThreadReplies = message?.threadInfo?.replyCount > 0 && replyType === 'THREAD';

  // onMouseDown: (e: React.MouseEvent<T>) => void;
  // onTouchStart: (e: React.TouchEvent<T>) => void;
  // onMouseUp: (e: React.MouseEvent<T>) => void;
  // onMouseLeave: (e: React.MouseEvent<T>) => void;
  // onTouchEnd: (e: React.TouchEvent<T>) => void;
  const longPress = useLongPress({
    onLongPress: () => {
      if (isMobile) {
        setShowMenu(true);
      }
    },
    onClick: () => {
      // @ts-ignore
      if (isMobile && isThumbnailMessage(message) && isSentMessage(message)) {
        showFileViewer(true);
      }
    },
  }, {
    delay: 300,
    shouldPreventDefault: false,
  });

  if (message?.isAdminMessage?.() || message?.messageType === 'admin') {
    return (<ClientAdminMessage message={message as AdminMessage} />);
  }
  return (
    <div
      className={getClassName([className, 'sendbird-message-content', isByMeClassName])}
      onMouseOver={() => setMouseHover(true)}
      onMouseLeave={() => setMouseHover(false)}
    >
      {/* left */}
      <div className={getClassName(['sendbird-message-content__left', isReactionEnabledClassName, isByMeClassName, useReplyingClassName])}>
        {(!isByMe && !chainBottom) && (
          /** user profile */
          <ContextMenu
            menuTrigger={(toggleDropdown: () => void): ReactElement => (
              <Avatar
                className={`sendbird-message-content__left__avatar ${displayThreadReplies ? 'use-thread-replies' : ''}`}
                // @ts-ignore
                src={channel?.members?.find((member) => member?.userId === message?.sender?.userId)?.profileUrl || message?.sender?.profileUrl || ''}
                // TODO: Divide getting profileUrl logic to utils
                ref={avatarRef}
                width="28px"
                height="28px"
                onClick={(): void => { if (!disableUserProfile) toggleDropdown(); }}
              />
            )}
            menuItems={(closeDropdown: () => void): ReactElement => (
              <MenuItems
                /**
                * parentRef: For catching location(x, y) of MenuItems
                * parentContainRef: For toggling more options(menus & reactions)
                */
                parentRef={avatarRef}
                parentContainRef={avatarRef}
                closeDropdown={closeDropdown}
                style={{ paddingTop: '0px', paddingBottom: '0px' }}
              >
                {renderUserProfile
                  // @ts-ignore
                  ? renderUserProfile({ user: message?.sender, close: closeDropdown })
                  // @ts-ignore
                  : (<UserProfile user={message.sender} onSuccess={closeDropdown} />)
                }
              </MenuItems>
            )}
          />
        )}
        {/* outgoing menu */}
        {isByMe && !isMobile && (
          <div className={getClassName(['sendbird-message-content-menu', isReactionEnabledClassName, supposedHoverClassName, isByMeClassName])}>
            <MessageItemMenu
              className="sendbird-message-content-menu__normal-menu"
              channel={channel}
              message={message as UserMessage | FileMessage}
              isByMe={isByMe}
              replyType={replyType}
              disabled={disabled}
              showEdit={showEdit}
              showRemove={showRemove}
              resendMessage={resendMessage}
              setQuoteMessage={setQuoteMessage}
              setSupposedHover={setSupposedHover}
              onReplyInThread={({ message }) => {
                if (threadReplySelectType === ThreadReplySelectType.THREAD) {
                  onReplyInThread({ message });
                } else if (threadReplySelectType === ThreadReplySelectType.PARENT) {
                  scrollToMessage(message.parentMessage.createdAt, message.parentMessageId);
                }
              }}
            />
            {isReactionEnabledInChannel && (
              <MessageItemReactionMenu
                className="sendbird-message-content-menu__reaction-menu"
                message={message as UserMessage | FileMessage}
                userId={userId}
                spaceFromTrigger={{}}
                emojiContainer={emojiContainer}
                toggleReaction={toggleReaction}
                setSupposedHover={setSupposedHover}
              />
            )}
          </div>
        )}
      </div>
      {/* middle */}
      <div
        className="sendbird-message-content__middle"
        {...(isMobile ? { ...longPress } : {})}
        ref={contentRef}
      >
        {(!isByMe && !chainTop && !useReplying) && (
          <Label
            className="sendbird-message-content__middle__sender-name"
            type={LabelTypography.CAPTION_2}
            color={LabelColors.ONBACKGROUND_2}
          >
            {
              // @ts-ignore
              channel?.members?.find((member) => member?.userId === message?.sender?.userId)?.nickname
              || getSenderName(message as UserMessage | FileMessage)
              // TODO: Divide getting profileUrl logic to utils
            }
          </Label>
        )}
        {/* quote message */}
        {(useReplying) ? (
          <div className={getClassName(['sendbird-message-content__middle__quote-message', isByMe ? 'outgoing' : 'incoming', useReplyingClassName])}>
            <QuoteMessage
              message={message as UserMessage | FileMessage}
              userId={userId}
              isByMe={isByMe}
              isUnavailable={(replyType === 'THREAD' && (channel?.joinedAt ?? 0 * 1000) > (message.parentMessage?.createdAt ?? 0))}
              onClick={() => {
                if (replyType === 'THREAD' && threadReplySelectType === ThreadReplySelectType.THREAD) {
                  onQuoteMessageClick?.({ message: message as UserMessage | FileMessage });
                }
                if (
                  (replyType === 'QUOTE_REPLY' || (replyType === 'THREAD' && threadReplySelectType === ThreadReplySelectType.PARENT))
                  && message?.parentMessage?.createdAt && message?.parentMessageId
                ) {
                  scrollToMessage(message.parentMessage.createdAt, message.parentMessageId);
                }
              }}
            />
          </div>
        ) : null}
        {/* container: message item body + emoji reactions */}
        <div className={getClassName(['sendbird-message-content__middle__body-container'])} >
          {/* message status component */}
          {(isByMe && !chainBottom) && (
            <div className={getClassName(['sendbird-message-content__middle__body-container__created-at', 'left', supposedHoverClassName])}>
              <div className="sendbird-message-content__middle__body-container__created-at__component-container">
                <MessageStatus
                  message={message as UserMessage | FileMessage}
                  channel={channel}
                />
              </div>
            </div>
          )}
          {/* message item body components */}
          {isTextMessage(message as UserMessage) && (
            <TextMessageItemBody
              className="sendbird-message-content__middle__message-item-body"
              message={message as UserMessage}
              isByMe={isByMe}
              mouseHover={mouseHover}
              isMentionEnabled={config?.isMentionEnabled || false}
              isReactionEnabled={isReactionEnabledInChannel}
            />
          )}
          {(isOGMessage(message as UserMessage)) && (
            <OGMessageItemBody
              className="sendbird-message-content__middle__message-item-body"
              message={message as UserMessage}
              isByMe={isByMe}
              mouseHover={mouseHover}
              isMentionEnabled={config?.isMentionEnabled || false}
              isReactionEnabled={isReactionEnabledInChannel}
              onMessageHeightChange={onMessageHeightChange}
            />
          )}
          {(getUIKitMessageType((message as FileMessage)) === messageTypes.FILE) && (
            <FileMessageItemBody
              className="sendbird-message-content__middle__message-item-body"
              message={message as FileMessage}
              isByMe={isByMe}
              mouseHover={mouseHover}
              isReactionEnabled={isReactionEnabledInChannel}
            />
          )}
          {isVoiceMessage(message as FileMessage) && (
            <VoiceMessageItemBody
              className="sendbird-message-content__middle__message-item-body"
              message={message as FileMessage}
              channelUrl={channel?.url ?? ''}
              isByMe={isByMe}
              isReactionEnabled={isReactionEnabledInChannel}
            />
          )}
          {(isThumbnailMessage(message as FileMessage)) && (
            <ThumbnailMessageItemBody
              className="sendbird-message-content__middle__message-item-body"
              message={message as FileMessage}
              isByMe={isByMe}
              mouseHover={mouseHover}
              isReactionEnabled={isReactionEnabledInChannel}
              showFileViewer={showFileViewer}
            />
          )}
          {(getUIKitMessageType((message as FileMessage)) === messageTypes.UNKNOWN) && (
            <UnknownMessageItemBody
              className="sendbird-message-content__middle__message-item-body"
              message={message}
              isByMe={isByMe}
              mouseHover={mouseHover}
              isReactionEnabled={isReactionEnabledInChannel}
            />
          )}
          {/* reactions */}
          {(isReactionEnabledInChannel && message?.reactions?.length > 0) && (
            <div className={getClassName([
              'sendbird-message-content-reactions',
              (!isByMe || isThumbnailMessage(message as FileMessage) || isOGMessage(message as UserMessage)) ? '' : 'primary',
              mouseHover ? 'mouse-hover' : '',
            ])}>
              <EmojiReactions
                userId={userId}
                message={message as UserMessage | FileMessage}
                channel={channel}
                isByMe={isByMe}
                emojiContainer={emojiContainer}
                memberNicknamesMap={nicknamesMap}
                toggleReaction={toggleReaction}
              />
            </div>
          )}
          {(!isByMe && !chainBottom) && (
            <Label
              className={getClassName(['sendbird-message-content__middle__body-container__created-at', 'right', supposedHoverClassName])}
              type={LabelTypography.CAPTION_3}
              color={LabelColors.ONBACKGROUND_2}
            >
              {format(message?.createdAt || 0, 'p', {
                locale: dateLocale,
              })}
            </Label>
          )}
        </div>
        {/* thread replies */}
        {displayThreadReplies && (
          <ThreadReplies
            className="sendbird-message-content__middle__thread-replies"
            threadInfo={message?.threadInfo}
            onClick={() => onReplyInThread?.({ message: message as UserMessage | FileMessage })}
          />
        )}
      </div>
      {/* right */}
      <div className={getClassName(['sendbird-message-content__right', chainTopClassName, isReactionEnabledClassName, useReplyingClassName])}>
        {/* incoming menu */}
        {!isByMe && !isMobile && (
          <div className={getClassName(['sendbird-message-content-menu', chainTopClassName, supposedHoverClassName, isByMeClassName])}>
            {isReactionEnabledInChannel && (
              <MessageItemReactionMenu
                className="sendbird-message-content-menu__reaction-menu"
                message={message as UserMessage | FileMessage}
                userId={userId}
                spaceFromTrigger={{}}
                emojiContainer={emojiContainer}
                toggleReaction={toggleReaction}
                setSupposedHover={setSupposedHover}
              />
            )}
            <MessageItemMenu
              className="sendbird-message-content-menu__normal-menu"
              channel={channel}
              message={message as UserMessage | FileMessage}
              isByMe={isByMe}
              replyType={replyType}
              disabled={disabled}
              showRemove={showRemove}
              resendMessage={resendMessage}
              setQuoteMessage={setQuoteMessage}
              setSupposedHover={setSupposedHover}
              onReplyInThread={({ message }) => {
                if (threadReplySelectType === ThreadReplySelectType.THREAD) {
                  onReplyInThread({ message });
                } else if (threadReplySelectType === ThreadReplySelectType.PARENT) {
                  scrollToMessage(message.parentMessage.createdAt, message.parentMessageId);
                }
              }}
            />
          </div>
        )}
      </div>
      {
        showMenu && (
          message?.isUserMessage?.() || message?.isFileMessage?.()
        ) && (
          <MobileMenu
            parentRef={contentRef}
            channel={channel}
            hideMenu={() => { setShowMenu(false); }}
            message={message}
            isReactionEnabled={isReactionEnabledInChannel}
            isByMe={isByMe}
            userId={userId}
            replyType={replyType}
            disabled={disabled}
            showRemove={showRemove}
            emojiContainer={emojiContainer}
            resendMessage={resendMessage}
            setQuoteMessage={setQuoteMessage}
            toggleReaction={toggleReaction}
            showEdit={showEdit}
            onReplyInThread={({ message }) => {
              if (threadReplySelectType === ThreadReplySelectType.THREAD) {
                onReplyInThread?.({ message });
              } else if (threadReplySelectType === ThreadReplySelectType.PARENT) {
                scrollToMessage?.(message?.parentMessage?.createdAt || 0, message?.parentMessageId || 0);
              }
            }}
          />
        )
      }
    </div>
  );
}
