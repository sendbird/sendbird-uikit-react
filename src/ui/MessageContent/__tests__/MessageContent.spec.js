import React from 'react';
import { mount } from 'enzyme';
import renderer from 'react-test-renderer';

import MessageContent from "../index";
jest.mock('date-fns/format', () => () => ('mock-date'));

const createMockChannel = (process) => {
  const mockChannel = {
    isGroupChannel: () => true,
    getUnreadMemberCount: () => 1,
    getUndeliveredMemberCount: () => 1,
  };
  return process ? process(mockChannel) : mockChannel;
};
const createMockMessage = (process) => {
  const mockMessage = {
    messageId: 1010,
    messageType: 'user',
    message: 'First second third',
    url: '',
    name: '',
    type: '',
    thumbnails: [],
    reactions: [],
    sendingStatus: 'succeeded',
    parentMessageId: 0,
    parentMessageInfo: null,
    sender: {
      profileUrl: '',
      userId: 'user-id-001',
      nickname: 'Mathew',
    },
    threadInfo: {
      replyCount: 0,
    },
    isAdminMessage: () => false,
    isUserMessage: () => true,
    isFileMessage: () => false,
    isResendable: () => false,
  };
  return process ? process(mockMessage) : mockMessage;
};

describe('MessageContent', () => {
  // TODO: Add tests after message threading is applied
  // useReplying <-> replyType
  // it('should render components by replyType', () => {});

  it('should have class names by own user basic status', () => {
    const component = mount(
      <MessageContent
        userId="sendbird-user-000"
        message={createMockMessage()}
        channel={createMockChannel()}
      />
    );
    expect(
      component.find('.sendbird-message-content').hostNodes().exists()
    ).toBe(true);
    expect(
      component.find('.sendbird-message-content__left').hostNodes().exists()
    ).toBe(true);
    expect(
      component.find('.sendbird-message-content__middle').hostNodes().exists()
    ).toBe(true);
    expect(
      component.find('.sendbird-message-content__right').hostNodes().exists()
    ).toBe(true);

    expect(
      component.find('.sendbird-message-content__left__avatar').hostNodes().exists()
    ).toBe(true);
    expect(
      component.find('.sendbird-message-content__middle__body-container__created-at.left').hostNodes().exists()
    ).toBe(false);
    expect(
      component.find('.sendbird-message-content-menu.outgoing').hostNodes().exists()
    ).toBe(false);
    expect(
      component.find('.sendbird-message-content__middle__sender-name').hostNodes().exists()
    ).toBe(true);
    expect(
      component.find('.sendbird-message-content-reactions').hostNodes().exists()
    ).toBe(false);
    expect(
      component.find('.sendbird-message-content__right.chain-top').hostNodes().exists()
    ).toBe(false);
    expect(
      component.find('.sendbird-message-content__middle__body-container__created-at.right').hostNodes().exists()
    ).toBe(true);
    expect(
      component.find('.sendbird-message-content-menu.incoming').hostNodes().exists()
    ).toBe(true);
    expect(
      component.find('.sendbird-message-content-menu.chain-top').hostNodes().exists()
    ).toBe(false);
    expect(
      component.find('.sendbird-message-content-menu.outgoing .sendbird-message-content-menu__normal-menu').hostNodes().exists()
    ).toBe(false);
    expect(
      component.find('.sendbird-message-content-menu.outgoing .sendbird-message-content-menu__reaction-menu').hostNodes().exists()
    ).toBe(false);
    expect(
      component.find('.sendbird-message-content-menu.incoming .sendbird-message-content-menu__normal-menu').hostNodes().exists()
    ).toBe(true);
    expect(
      component.find('.sendbird-message-content-menu.incoming .sendbird-message-content-menu__reaction-menu').hostNodes().exists()
    ).toBe(false);
  });

  it('should render components by isByMe prop', () => {
    const outgoingMessage = mount(
      <MessageContent
        userId="user-id-001"
        message={createMockMessage()}
        channel={createMockChannel()}
        isByMe
      />
    );
    expect(
      outgoingMessage.find('.sendbird-message-content.outgoing').hostNodes().exists()
    ).toBe(true);
    expect(
      outgoingMessage.find('.sendbird-message-content__left.outgoing').hostNodes().exists()
    ).toBe(true);
    expect(
      outgoingMessage.find('.sendbird-message-content-menu.outgoing').hostNodes().exists()
    ).toBe(true);
    expect(
      outgoingMessage.find('.sendbird-message-content.incoming').hostNodes().exists()
    ).toBe(false);
    expect(
      outgoingMessage.find('.sendbird-message-content__left.incoming').hostNodes().exists()
    ).toBe(false);
    expect(
      outgoingMessage.find('.sendbird-message-content-menu.incoming').hostNodes().exists()
    ).toBe(false);
    const incomingMessage = mount(
      <MessageContent
        userId="user-id-002"
        message={createMockMessage()}
        channel={createMockChannel()}
        isByMe={false}
      />
    );
    expect(
      incomingMessage.find('.sendbird-message-content.outgoing').hostNodes().exists()
    ).toBe(false);
    expect(
      incomingMessage.find('.sendbird-message-content__left.outgoing').hostNodes().exists()
    ).toBe(false);
    expect(
      incomingMessage.find('.sendbird-message-content-menu.outgoing').hostNodes().exists()
    ).toBe(false);
    expect(
      incomingMessage.find('.sendbird-message-content.incoming').hostNodes().exists()
    ).toBe(true);
    expect(
      incomingMessage.find('.sendbird-message-content__left.incoming').hostNodes().exists()
    ).toBe(true);
    expect(
      incomingMessage.find('.sendbird-message-content-menu.incoming').hostNodes().exists()
    ).toBe(true);
  });

  it('should not render components when chainTop is true', () => {
    const component = mount(
      <MessageContent
        userId="sendbird-user-000"
        message={createMockMessage()}
        channel={createMockChannel()}
        chainTop
      />
    );
    expect(
      component.find('.sendbird-message-content__middle__sender-name').hostNodes().exists()
    ).toBe(false);
    expect(
      component.find('.sendbird-message-content__right.chain-top').hostNodes().exists()
    ).toBe(true);
    expect(
      component.find('.sendbird-message-content-menu.chain-top').hostNodes().exists()
    ).toBe(true);
  });

  it('should not render components when chainBottom is true', () => {
    const outgoingMessage = mount(
      <MessageContent
        userId="sendbird-user-001"
        message={createMockMessage()}
        channel={createMockChannel()}
        isByMe
        chainBottom
      />
    );
    expect(
      outgoingMessage.find('.sendbird-message-content__left__avatar').hostNodes().exists()
    ).toBe(false);
    expect(
      outgoingMessage.find('.sendbird-message-content__middle__body-container__created-at').hostNodes().exists()
    ).toBe(false);
    const incomingMessage = mount(
      <MessageContent
        userId="sendbird-user-002"
        message={createMockMessage()}
        channel={createMockChannel()}
        isByMe={false}
        chainBottom
      />
    );
    expect(
      incomingMessage.find('.sendbird-message-content__left__avatar').hostNodes().exists()
    ).toBe(false);
    expect(
      incomingMessage.find('.sendbird-message-content__middle__body-container__created-at').hostNodes().exists()
    ).toBe(false);
  });

  // it('should render components by useReaction and reactions', () => {});

  it('should do a snapshot test of the MessageContent DOM', function () {
    const component = renderer.create(
      <MessageContent
        className="classname-for-snapshot"
        message={createMockMessage()}
        channel={createMockChannel()}
        userId="user-id-000"
      />,
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
