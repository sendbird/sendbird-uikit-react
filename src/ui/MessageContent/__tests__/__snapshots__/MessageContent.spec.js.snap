// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ui/MessageContent should do a snapshot test of the MessageContent DOM 1`] = `
<DocumentFragment>
  <div
    class="classname-for-snapshot sendbird-message-content incoming"
  >
    <div
      class="sendbird-message-content__left  incoming "
    >
      <div
        class="sendbird-context-menu"
        style="display: inline;"
      >
        <div
          class="sendbird-message-content__left__avatar  sendbird-avatar"
          role="button"
          style="height: 28px; width: 28px; z-index: 0; left: 0px;"
          tabindex="0"
        >
          <div
            class="sendbird-avatar-img sendbird-image-renderer"
            style="width: 100%; min-width: min(400px, 28px); max-width: 400px;"
          >
            <div
              class="sendbird-image-renderer__image"
              style="width: 100%; min-width: min(400px, 28px); max-width: 400px; position: absolute; background-repeat: no-repeat; background-position: center; background-size: cover; background-image: url();"
            />
            <img
              alt=""
              class="sendbird-image-renderer__hidden-image-loader"
              src=""
            />
          </div>
        </div>
      </div>
    </div>
    <div
      class="sendbird-message-content__middle"
    >
      <span
        class="sendbird-message-content__middle__sender-name sendbird-label sendbird-label--caption-2 sendbird-label--color-onbackground-2"
      >
        Mathew
      </span>
      <div
        class="sendbird-message-content__middle__body-container"
      >
        <span
          class="sendbird-label sendbird-label--body-1 sendbird-label--color-onbackground-1"
        >
          <div
            class="sendbird-message-content__middle__message-item-body sendbird-text-message-item-body incoming  "
          >
            First second third
          </div>
        </span>
        <span
          class="sendbird-message-content__middle__body-container__created-at right  sendbird-label sendbird-label--caption-3 sendbird-label--color-onbackground-2"
        >
          mock-date
        </span>
      </div>
    </div>
    <div
      class="sendbird-message-content__right   "
    >
      <div
        class="sendbird-message-content-menu   incoming"
      >
        <div
          class="sendbird-message-content-menu__normal-menu sendbird-message-item-menu"
        >
          <div
            class="sendbird-context-menu"
            style="display: inline;"
          >
            <button
              class="sendbird-message-item-menu__trigger sendbird-iconbutton "
              style="height: 32px; width: 32px;"
              type="button"
            >
              <span
                class="sendbird-iconbutton__inner"
              >
                <div
                  class="sendbird-message-item-menu__trigger__icon sendbird-icon sendbird-icon-more sendbird-icon-color--content-inverse"
                  role="button"
                  style="width: 24px; min-width: 24px; height: 24px; min-height: 24px;"
                  tabindex="0"
                >
                  <test-file-stub />
                </div>
              </span>
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
